lib/system/freertos/xlnx_common/zynqmp_aarch64/sys.c

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (c) 2023 Advanced Micro Devices, Inc. All Rights Reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+/*
+ * @file	freertos/xlnx_common/zynqmp_aarch64/sys.c
+ * @brief	machine specific system primitives implementation.
+ */
+
+#include <metal/compiler.h>
+#include <metal/io.h>
+#include <metal/sys.h>
+#include <metal/utilities.h>
+#include <stdint.h>
+#include "xil_cache.h"
+#include "xil_exception.h"
+#include "xil_mmu.h"
+#include "xscugic.h"
+
+/* System Device Tree (SDT) flow does not have the files generated. */
+#ifndef SDT
+
+#ifdef VERSAL_NET
+#include "xcpu_cortexa78.h"
+#elif defined(versal)
+#include "xcpu_cortexa72.h"
+#else
+#include "xreg_cortexa53.h"
+#endif /* defined(versal) */
+
+#endif /* !SDT */
+
+void sys_irq_restore_enable(unsigned int flags)
+{
+	Xil_ExceptionEnableMask(~flags);
+}
+
+unsigned int sys_irq_save_disable(void)
+{
+	unsigned int state = mfcpsr() & XIL_EXCEPTION_ALL;
+
+	if (state != XIL_EXCEPTION_ALL)
+		Xil_ExceptionDisableMask(XIL_EXCEPTION_ALL);
+
+	return state;
+}
+
+void metal_machine_cache_flush(void *addr, unsigned int len)
+{
+	if (!addr || !len)
+		Xil_DCacheFlush();
+	else
+		Xil_DCacheFlushRange((intptr_t)addr, len);
+}
+
+void metal_machine_cache_invalidate(void *addr, unsigned int len)
+{
+	if (!addr || !len)
+		Xil_DCacheInvalidate();
+	else
+		Xil_DCacheInvalidateRange((intptr_t)addr, len);
+}
+
+/**
+ * @brief poll function until some event happens
+ */
+inline void metal_weak metal_freertos_default_poll(void)
+{
+	metal_asm volatile("wfi");
+}
+
+/* TODO: document functionality of this function. */
+void *metal_machine_io_mem_map(void *va, metal_phys_addr_t pa,
+			       size_t size, unsigned int flags)
+{
+	unsigned long section_offset;
```

Comment:
I prefer we don' have platform specific implementation in libmetal. Instead push this implementation in Xilinx's BSP and use only interfaces here from the BSP. That way, It's clean transformation from libmetal to BSP.

I see Xil_MemMap API is available to achieve mapping of memory for ZynqMP R5 platform so, Probably that API needs to be ported for other Xilinx Platforms' BSP.